### PR TITLE
Update category

### PR DESCRIPTION
**Pull request by chubi-x**
finished route to update a category (including adding tweets) @gmemmy

### CODE REVIEW

File: routes/routes.ts [hunk 1]
```diff
@@ -284,6 +284,55 @@ router.post("/category", async (req: Request, res: Response) => {
...
+  try {
+    // user has to have a session
+    if (req.session.userId) {
+      const userId = req.session.userId;
+      const categoryId = req.params.id;
+      const categoryRef = usersRef
+        .child(userId)
+        .child("categories")
+        .child(categoryId);
+      await categoryRef.update(
```
Comment (gmemmy) [inline routes/routes.ts:297]: I think we have several ways of updating a category;
- Adding a bookmark to it
- Deleting a bookmark from it
- Changing individual attributes like the category name, description, etc.

We need a way to specify what type of update the client is trying to make, then use that type to determine what gets updated in the firebase doc.

We could use this route for handling updates(add, delete) to the categories bookmarks. Maybe the request body could have a property with two possible values of add or delete, then we first fetch the doc for that category, append to it and update the doc if we are adding a new bookmark or remove a bookmark with the provided bookmark id from our request body. This means each bookmark should have an id, if it already doesn't have.

Comment (gmemmy) [reply routes/routes.ts:297]: We can create a separate route for updating the category attributes(name, description, etc).

**Review by gmemmy — CHANGES_REQUESTED** [review]

File: routes/routes.ts (hunk 1 above)
Comment (chubi-x) [reply routes/routes.ts:297]: > I think we have several ways of updating a category;
> 
> * Adding a bookmark to it
> * Deleting a bookmark from it
> * Changing individual attributes like the category name, description, etc.
> 
> We need a way to specify what type of update the client is trying to make, then use that type to determine what gets updated in the firebase doc.
> 
> We could use this route for handling updates(add, delete) to the categories bookmarks. Maybe the request body could have a property with two possible values of add or delete, then we first fetch the doc for that category, append to it and update the doc if we are adding a new bookmark or remove a bookmark with the provided bookmark id from our request body. This means each bookmark should have an id, if it already doesn't have.

This makes sense. the req body can have a 'update_type' field (which could be an enum with two possible values), then based on that field, I'll update the category accordingly. I'll get on it now. Thanks for your help.

Comment (chubi-x) [reply routes/routes.ts:297]: > We can create a separate route for updating the category attributes(name, description, etc).

alright. will do.

Comment (gmemmy) [reply routes/routes.ts:297]: Yes, this works!

**Comment by chubi-x** [comment]
How to test @gmemmy  :

### Update category
1. send a patch request to /category/<category id>/bookmarks
2.. payload should contain :
            updateType - string that specifies the action to take. can be 'add' or 'delete'
            bookmarks: array of strings that contains the tweet ids of bookmarks to update
            
            - Example payload: 
            ```
            {
              "updateType": "add",
              "bookmarks": ["000", "123", "456"]
            }
            ```

### Update category attributes
1. send a patch request to /category/<category id>
2. example payload:
     ```
       {
              "name":"threejs",
              "description": "threejs bookmarks",
              "image":"https://threejs.com"
       }
```

File: routes/routes.ts [hunk 2]
```diff
@@ -283,7 +292,196 @@ router.post("/category", async (req: Request, res: Response) => {
...
+                          return res.send(
+                            "There was an error removing this bookmark. please try again."
+                          );
+                        }
+                      });
+                    }
+                  });
+                });
+                // return the updated category
+                return await returnCategory();
```
Comment (gmemmy) [inline routes/routes.ts:451]: We don't need to return the category from this endpoint.

**Review by gmemmy — CHANGES_REQUESTED** [review]

**Pull request merged**